src/computability/list_blank.lean

```diff
@@ -0,0 +1,635 @@
+  list_val.head_map, list_val.modify_nth, list_val.map_cons, list_val.tail_map]
+
+/-- Append a list on the left side of a list_val. -/
+@[simp] def list_val.append {Γ} {val : Γ} : list Γ → list_val Γ val → list_val Γ val
+| [] L := L
+| (a :: l) L := list_val.cons a (list_val.append l L)
+
+@[simp] theorem list_val.append_mk {Γ} (val : Γ) (l₁ l₂ : list Γ) :
+  list_val.append l₁ (list_val.mk val l₂) = list_val.mk val (l₁ ++ l₂) :=
+by induction l₁; simp only [*,
+     list_val.append, list.nil_append, list.cons_append, list_val.cons_mk]
+
+theorem list_val.append_assoc {Γ} {val : Γ} (l₁ l₂ : list Γ) (l₃ : list_val Γ val) :
+  list_val.append (l₁ ++ l₂) l₃ = list_val.append l₁ (list_val.append l₂ l₃) :=
+l₃.induction_on $ by intro; simp only [list_val.append_mk, list.append_assoc]
+
+/-- The `bind` function on lists is well defined on `list_val`s provided that the default element
+is sent to a sequence of default elements. -/
+def list_val.bind {Γ Γ'} {val : Γ} (val' : Γ')
+  (l : list_val Γ val) (f : Γ → list Γ')
+  (hf : ∃ n, f val = list.repeat val' n) : list_val Γ' val' :=
+l.lift_on (λ l, list_val.mk val' (list.bind l f)) begin
+  rintro l _ ⟨i, rfl⟩, cases hf with n e, refine quotient.sound' (or.inl ⟨i * n, _⟩),
+  rw [list.bind_append, mul_comm], congr,
+  induction i with i IH, refl,
+  simp only [IH, e, list.repeat_add, nat.mul_succ, add_comm, list.repeat_succ, list.cons_bind],
+end
+
+@[simp] lemma list_val.bind_mk {Γ Γ'} (val : Γ) (val' : Γ')
+  (l : list Γ) (f : Γ → list Γ') (hf) :
+  (list_val.mk val l).bind val' f hf = list_val.mk val' (l.bind f) := rfl
+
+@[simp] lemma list_val.cons_bind {Γ Γ'} {val : Γ} {val' : Γ'}
+  (a : Γ) (l : list_val Γ val) (f : Γ → list Γ') (hf) :
+  (l.cons a).bind val' f hf = (l.bind val' f hf).append (f a) :=
+l.induction_on $ by intro; simp only [list_val.append_mk,
+  list_val.bind_mk, list_val.cons_mk, list.cons_bind]
+
+
+section list_blank
+
+/-- A `list_blank Γ` is a quotient of `list Γ` by extension by blanks at the end. This is used to
+represent half-tapes of a Turing machine, so that we can pretend that the list continues
+infinitely with blanks. -/
+def list_blank (Γ) [inhabited Γ] := list_val Γ default
+
+instance list_blank.inhabited {Γ} [inhabited Γ] : inhabited (list_blank Γ) := ⟨quotient.mk' []⟩
+instance list_blank.has_emptyc {Γ} [inhabited Γ] : has_emptyc (list_blank Γ) := ⟨quotient.mk' []⟩
+
+/-- The quotient map turning a `list` into a `list_blank`. -/
+def list_blank.mk {Γ} [inhabited Γ] : list Γ → list_blank Γ := quotient.mk'
+
+@[elab_as_eliminator]
+protected lemma list_blank.induction_on {Γ} [inhabited Γ]
+  {p : list_blank Γ → Prop} (q : list_blank Γ)
+  (h : ∀ a, p (list_blank.mk a)) : p q := quotient.induction_on' q h
+
+/-- The head of a `list_blank` is well defined. -/
+def list_blank.head {Γ} [inhabited Γ] (l : list_blank Γ) : Γ :=
+l.lift_on list.head begin
+  rintro _ _ ⟨i, rfl⟩,
+  cases a, {cases i; refl}, refl
+end
```

Comment:
This can also be
```suggestion
list_val.head l
```
right? Any reason for redoing the definition? (And the same for below.)

Reply:
Ok Done.

Reply:
This seems to have screwed up a lot of the proofs that use this API, BTW, not sure if it was worth it.